### PR TITLE
[Trainer] Implement GRPOTrainer and SOTA execution pipelines

### PR DESCRIPTION
**Pull request by coder-jayp**
## Description

This PR introduces `GRPOTrainer` into the core `torchrl/trainers/algorithms` suite, bringing native support for RLHF (Reinforcement Learning from Human Feedback) architectures into TorchRL. 

Key architectural implementations include:
- **`GRPOTrainer`**: Integrated natively with the standard `Trainer` class to reuse existing data collection and hook logic.
- **`GRPOOptimizationStepper`**: Decouples `autocast`, gradient accumulation, and gradient clipping logic from the main loop.
- **`WeightSyncHook`**: Provides robust, asynchronous weight synchronization tailored for distributed LLM inference engines (e.g., vLLM, sglang), ensuring models stay in sync at precise macro-steps.
- **Hydra Config Parity**: Introduced `GRPOTrainerConfig` mapped 1:1 with the initializer to ensure non-breaking support for Hydra configuration pipelines.
- **SOTA Refactor & CI**: Refactored `sota-implementations/grpo/` (both sync and async) to utilize this new abstracted API, and added them to the GitHub Actions smoke test matrix via `sota-check`.
- **Testing**: Added 23 comprehensive unit tests validating accumulation boundaries, sync triggers, and extreme edge cases.

## Motivation and Context

As TorchRL continues to expand its utility for Large Language Model tuning (RLHF/PPO/GRPO), the lack of a standardized, abstracted `GRPOTrainer` meant that implementations required massive, monolithic custom training loops (as previously seen in `sota-implementations/grpo`). 

This PR solves this by providing a unified, modular, and heavily tested Trainer API specific to GRPO. It massively simplifies the code required by end-users to train LLMs via GRPO while enabling seamless integration with external high-throughput inference engines like vLLM.

- [ ] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [x] New feature (non-breaking change which adds core functionality)
- [x] Documentation (update in the documentation)
- [x] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by coder-jayp** [comment]
@torchrlbot reviewer @vmoens

**Comment by vmoens** [comment]
I replaced the large set of tests that inspected private fields, closures, exports, and constructor wiring with a compact behavior-focused suite.
The new tests exercise the public training path: policy updates, gradient-accumulation equivalence, weight synchronization at actual optimizer-step boundaries, metric logging, collector shutdown, and meaningful validation errors.

This reduces the GRPO test addition from 544 to 166 lines while providing stronger regression coverage. Exercising the feature end to end also exposed two real hook-contract bugs in weight synchronization and KL logging, which this commit fixes.

**Review by vmoens — APPROVED** [review]
LGTM thanks!

**Comment by coder-jayp** [comment]
@vmoens What would you prefer I pick up next? 
I'd be happy to work on a Trainer for DiffusionBC, DreamerV3, or whatever highest priority right now.

**Comment by theap06** [comment]
@coder-jayp I think features and trainer coverage have been sufficiently been implemented for the v0.14 release. Thanks! If you want to help around, I think many of the current tests need to be updated for github workflows. Additionally, we haven't been able to showcase any tutorials with results for the VLA/LeRobot/TRL implementations.

**Comment by coder-jayp** [comment]
Thanks @theap06. I'll put together a script for tutorials/sphinx-tutorials/ that showcases training an LLM using the TRL integration and produces actual learning curves/results.

File: torchrl/trainers/algorithms/grpo.py [hunk 1]
```diff
@@ -0,0 +1,526 @@
...
+        self._optimizer_step_count: int = 0
+
+    # ------------------------------------------------------------------
+    # Checkpointing (optimizer + scaler state)
+    # ------------------------------------------------------------------
+
+    def state_dict(self) -> dict[str, Any]:
+        sd: dict[str, Any] = {
+            "optimizer": self.optimizer.state_dict(),
+            "micro_step": self._micro_step,
```
Comment (gtnv) [inline torchrl/trainers/algorithms/grpo.py]: restoring micro_step=1 does not restore the partial gradients. my two-microbatch result changed from 1.0 to 1.1. can we save those gradients or reject mid-window checkpoints?

File: torchrl/trainers/algorithms/grpo.py [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+                ]
+                grad_norm = float(nn.utils.clip_grad_norm_(params, self.clip_norm))
+
+            if self._use_scaler:
+                self.scaler.step(self.optimizer)
+                self.scaler.update()
+            else:
+                self.optimizer.step()
+
+            self._optimizer_step_count += 1
```
Comment (gtnv) [inline torchrl/trainers/algorithms/grpo.py]: an overflow can skip GradScaler.step(). _optimizer_step_count still increments here. can it move only after a real parameter update?

File: torchrl/trainers/algorithms/grpo.py (hunk 1 above)
Comment (coder-jayp) [reply torchrl/trainers/algorithms/grpo.py]: Saving gradients bloats the checkpoint. state_dict() now throws a RuntimeError if called mid-accumulation.

File: torchrl/trainers/algorithms/grpo.py (hunk 2 above)
Comment (coder-jayp) [reply torchrl/trainers/algorithms/grpo.py]: Fixed. We now compare scaler.get_scale() before and after scaler.update() and bypass the increment if the scale decreases.

**Comment by vmoens** [comment]
@gtnv @coder-jayp I went on and merged this. It would be nice to have a script in sota-implementations that uses this and to show that the reward curve goes up for various tasks like instruction following, math, gsm8k, etc

**Comment by coder-jayp** [comment]
@vmoens Since I already wrote grpo-sync.py in this PR, I’ll run it on a small model to generate a proof-of-concept reward curve and open a follow-up PR with the plots.

**Comment by coder-jayp** [comment]
@vmoens Since NCCL strictly doesn't allow mapping both the Trainer and vLLM to the same device, I can't test it locally. Could you run sota-implementations/grpo/grpo-sync.py to generate the reward curves?
